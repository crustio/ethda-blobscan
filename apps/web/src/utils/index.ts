export * from "./charts";
export * from "./ethereum";
export * from "./date";
export * from "./explorers";
export * from "./routes";
export * from "./search";
export * from "./number";
export * from "./text";
export * from "./tra";
export const hiddenGasElement = true;
