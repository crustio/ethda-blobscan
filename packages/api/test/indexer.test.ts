import { TRPCError } from "@trpc/server";
import type { inferProcedureInput } from "@trpc/server";
import type { NodeHTTPRequest } from "@trpc/server/adapters/node-http";
import jwt from "jsonwebtoken";
import { describe, expect, it, vi } from "vitest";

import type { TRPCContext } from "../src/context";
import { createTRPCInnerContext, getJWTFromRequest } from "../src/context";
import { appRouter } from "../src/root";
import type { AppRouter } from "../src/root";

vi.mock("../src/env", () => ({
  env: {
    SECRET_KEY: "supersecret",
  },
}));

describe("Indexer router", async () => {
  it("should return the current slot", async () => {
    const ctx = (await createTRPCInnerContext()) as TRPCContext;
    const caller = appRouter.createCaller(ctx);

    const result = await caller.indexer.getSlot();

    expect(result).toMatchObject({ slot: 0 });
  });

  it("should not update the slot if not auth", async () => {
    const ctx = (await createTRPCInnerContext()) as TRPCContext;
    const caller = appRouter.createCaller(ctx);
    type Input = inferProcedureInput<AppRouter["indexer"]["updateSlot"]>;
    const input: Input = {
      slot: 10,
    };

    await expect(caller.indexer.updateSlot(input)).rejects.toThrow(
      new TRPCError({ code: "UNAUTHORIZED" })
    );
  });

  it("should update the slot if auth", async () => {
    const token = jwt.sign("foobar", "supersecret");
    const req = {
      headers: { authorization: `Bearer ${token}` },
    } as NodeHTTPRequest;
    const apiClient = getJWTFromRequest(req);
    console.log("ApiClient", apiClient);
    const ctx = (await createTRPCInnerContext({
      apiClient: apiClient,
    })) as TRPCContext;

    const caller = appRouter.createCaller(ctx);
    type Input = inferProcedureInput<AppRouter["indexer"]["updateSlot"]>;
    const input: Input = {
      slot: 10,
    };

    await caller.indexer.updateSlot(input);

    const result = await caller.indexer.getSlot();

    expect(result).toMatchObject({ slot: 10 });
  });

  it("should index new data if auth", async () => {
    const token = jwt.sign("foobar", "supersecret");
    const req = {
      headers: { authorization: `Bearer ${token}` },
    } as NodeHTTPRequest;
    const apiClient = getJWTFromRequest(req);
    console.log("ApiClient", apiClient);
    const ctx = (await createTRPCInnerContext({
      apiClient: apiClient,
    })) as TRPCContext;

    const caller = appRouter.createCaller(ctx);
    type Input = inferProcedureInput<AppRouter["indexer"]["indexData"]>;
    const input: Input = {
      block: {
        number: 1003,
        hash: "blockHash001",
        timestamp: 1630257174,
        slot: 103,
        blobGasUsed: "10000",
        excessBlobGas: "5000",
      },
      transactions: [
        {
          hash: "txHash008",
          from: "address1",
          to: "address3",
          blockNumber: 1003,
          gasPrice: "20000",
          maxFeePerBlobGas: "15000",
        },
        {
          hash: "txHash009",
          from: "address4",
          to: "address6",
          blockNumber: 1003,
          gasPrice: "10000",
          maxFeePerBlobGas: "1800",
        },
        {
          hash: "txHash010",
          from: "address4",
          to: "address2",
          blockNumber: 1003,
          gasPrice: "3000000",
          maxFeePerBlobGas: "20000",
        },
      ],
      blobs: [
        {
          versionedHash: "blobHash008",
          commitment: "commitment008",
          data: "0x34567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
          txHash: "txHash008",
          index: 0,
        },
        {
          versionedHash: "blobHash009",
          commitment: "commitment009",
          data: "0x34567890abcdef1234567890abcdef",
          txHash: "txHash009",
          index: 0,
        },
        {
          versionedHash: "blobHash010",
          commitment: "commitment010",
          data: "0x34567890abcdef1234567890abcdef1234567890abcdef",
          txHash: "txHash010",
          index: 0,
        },
      ],
    };

    await caller.indexer.indexData(input);

    const result = await caller.block.getByBlockNumber({
      number: 1003,
    });
    expect(result).toMatchSnapshot();
  });
});
