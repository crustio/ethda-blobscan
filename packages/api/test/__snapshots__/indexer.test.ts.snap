// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Indexer router > should index new data if auth 1`] = `
{
  "blobAsCalldataGasUsed": 1104n,
  "blobGasPrice": 1n,
  "blobGasUsed": 10000n,
  "excessBlobGas": 5000n,
  "hash": "blockHash001",
  "number": 1003,
  "slot": 103,
  "timestamp": 2021-08-29T17:12:54.000Z,
  "transactions": [
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
      ],
      "fromId": "address1",
      "hash": "txHash001",
      "toId": "address2",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 0,
        },
      ],
      "fromId": "address3",
      "hash": "txHash002",
      "toId": "address4",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1200,
          },
          "blobHash": "blobHash003",
          "index": 0,
        },
      ],
      "fromId": "address5",
      "hash": "txHash003",
      "toId": "address6",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 31,
          },
          "blobHash": "blobHash008",
          "index": 0,
        },
      ],
      "fromId": "address1",
      "hash": "txHash008",
      "toId": "address3",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 15,
          },
          "blobHash": "blobHash009",
          "index": 0,
        },
      ],
      "fromId": "address4",
      "hash": "txHash009",
      "toId": "address6",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 23,
          },
          "blobHash": "blobHash010",
          "index": 0,
        },
      ],
      "fromId": "address4",
      "hash": "txHash010",
      "toId": "address2",
    },
  ],
}
`;
